File: android/app/src/main/kotlin/com/fatwabot/app/notifications/ContentReminderPreferenceStore.kt
package com.fatwabot.app.notifications

import android.content.Context
import com.fatwabot.core.content.ContentReminderPreferences

/**
 * Persistence for the daily azkar/hadith reminder settings.
 *
 * Deliberately a separate SharedPreferences file from `notification_preferences`
 * (the prayer settings) — the two are edited independently, and sharing a file
 * would risk one feature's migration clearing the other's keys. Flat keys rather
 * than a serialized blob, matching `NotificationPreferenceStore`'s house style.
 */
class ContentReminderPreferenceStore(private val context: Context) {

    fun load(): ContentReminderPreferences {
        val prefs = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
        val defaults = ContentReminderPreferences()
        return ContentReminderPreferences(
            enabled = prefs.getBoolean(KEY_ENABLED, defaults.enabled),
            perDay = ContentReminderPreferences.clamp(prefs.getInt(KEY_PER_DAY, defaults.perDay)),
        )
    }

    fun save(preferences: ContentReminderPreferences) {
        context.getSharedPreferences(PREFS, Context.MODE_PRIVATE).edit()
            .putBoolean(KEY_ENABLED, preferences.enabled)
            .putInt(KEY_PER_DAY, ContentReminderPreferences.clamp(preferences.perDay))
            .apply()
    }

    private companion object {
        const val PREFS = "content_reminder_preferences"
        const val KEY_ENABLED = "enabled"
        const val KEY_PER_DAY = "per_day"
    }
}

File: ios/App/Sources/Notifications/ContentReminderPreferenceStore.swift
import ContentKit
import Foundation

/// Persistence for the daily azkar/hadith reminder settings. Deliberately a
/// separate blob from `notifications.prayer.preferences.v1` — the two are edited
/// independently, and sharing a key would mean a decode failure in one wiped the
/// other.
protocol ContentReminderPreferenceStoring: Sendable {
    func load() -> ContentReminderPreferences
    func save(_ preferences: ContentReminderPreferences)
}

final class UserDefaultsContentReminderPreferenceStore: ContentReminderPreferenceStoring, @unchecked Sendable {
    private static let key = "notifications.content.preferences.v1"
    private let defaults: UserDefaults

    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
    }

    /// Falls back to first-run defaults (on, 2/day) rather than throwing: a
    /// corrupt blob should cost the user their customisation, not their reminders.
    func load() -> ContentReminderPreferences {
        guard let data = defaults.data(forKey: Self.key),
              let decoded = try? JSONDecoder().decode(ContentReminderPreferences.self, from: data)
        else { return ContentReminderPreferences() }
        return decoded
    }

    func save(_ preferences: ContentReminderPreferences) {
        guard let data = try? JSONEncoder().encode(preferences) else { return }
        defaults.set(data, forKey: Self.key)
    }
}
